### ISSUE TITLE
Test `InferRecFunWithConstrainedArgument.purs` contains an unreliable comparison

### ISSUE THREAD

**Issue by andyarvanitis** [purescript/purescript#2093]
Compiler test:
`examples/passing/InferRecFunWithConstrainedArgument.purs`

```PureScript
module Main where
import Prelude

test 100.0 = 100.0
test n = test(1.0 + n)

main = Control.Monad.Eff.Console.print $ test 0.0
```

uses a floating-point number equality comparison (following some computations). This is unreliable and unsafe. In this case, you'll occasionally (somewhat randomly) get an infinite loop.

Discovered with the C++ backend, but same issue applies to JavaScript. I think the test should be changed, but maybe the comparison operators for `Number` (floating point) should be discussed?

**Comment by hdgarrood** [comment]
I guess it would make more sense to use Int in this test, but since we're using whole numbers here I think we ought to be okay anyway, at least in JS, as 0-100 are all safe integers: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger

What is the underlying representation for `Number` in C++ again?

**Comment by hdgarrood** [comment]
Oh, actually, I suppose it's not safe to assume that if x, y, and z can be represented exactly, and x + y = z (in reality), then x + y = z (with floating point), is it...

Are you suggesting an Eq Number instance which uses a delta and approximate comparisons? I think I'd rather keep transitivity (which I think we do have now, at least).

**Comment by AppShipIt** [comment]
> Are you suggesting an Eq Number instance which uses a delta and approximate comparisons?

That was my first thought but I suppose there'd also be a slight performance penalty.

> I think I'd rather keep transitivity (which I think we do have now, at least).

Reading a [post comparing](https://codewords.recurse.com/issues/one/when-is-equality-transitive-and-other-floating-point-curiosities) this issue in different languages.

**Comment by garyb** [comment]
If we were going to start making equality for floats a bit fuzzy we may as well just go whole hog and use arbitrary precision default numeric types.

I was thinking about this the other day actually - I wonder how bad it would be in terms of performance to do so, as at least then we'd have law abiding numbers by default. I'm thinking we could have them correct by default, but with the option to change type and use the much better performing yet entirely broken (from a laws point of view) number types where it matters. I guess we'd want overloaded literals for that to work too though. It also goes a little against the "no runtime" thing we have going on. It just bothers me that two of the most commonly used types behave really badly.

Anyway, aside from all that I agree that using `Int` may be a better fit for this test. :wink:

**Comment by hdgarrood** [comment]
Do we have any libraries offering a law-abiding real-number type which supports everything that `Number` does? There's [purescript-hugenums](http://pursuit.purescript.org/packages/purescript-hugenums), for instance, but that doesn't support division (and I can't think of a way it could do so without some kind of rounding).

Regarding the test, I'm also a bit confused that I've never seen this infinite loop happen with the JS backend. We've been using Travis for a long time, and that test has been there since Feb 2014 (!). I think it's probably safe to say that it's fine in JS. Perhaps it's worth investigating why JS and C++ are different here.

I am personally happy with the default being native JS numbers. Numeric inaccuracies are a pain but at least there's plenty of information about dealing with them. Also, I think switching away from native numbers would really hurt interop (which is supposed to be one of PureScript's main strengths, right?). Not only would it be annoying when working with JS code having to convert back and forth all the time, it would also be super easy to forget to do this when defining bindings using the FFI, causing runtime errors.

**Comment by AppShipIt** [comment]
> Perhaps it's worth investigating why JS and C++ are different here.

Could it be `float` versus `double` ([post](http://stackoverflow.com/a/16595884) with [example](http://codepad.org/c5xVNVDA))?

**Comment by hdgarrood** [comment]
According to https://github.com/andyarvanitis/purescript-native#other-notes, the C++ backend uses `double` for `Number`.

**Comment by hdgarrood** [comment]
Oh, also: floating point certainly has gotchas, but is it really indeterministic when you're just adding numbers together? That is, are you sure that floating point inaccuracy is the cause here? Surely such a basic operation as addition would be fairly well nailed down in the spec?

**Comment by AppShipIt** [comment]
> is it really indeterministic when you're just adding numbers together?

JavaScript no (I agree with your conclusion about keeping it native), C++ `double` no, but `float` apparently (found several posts beyond my experience level).

**Comment by jplatte** [comment]
https://en.wikipedia.org/wiki/Floating_point#Accuracy_problems

**Comment by jplatte** [comment]
TL;DR: It's definitely a problem in both JS and C++, and with both data types. With double you just usually don't notice, as your number need to be quite a few orders of magnitude different from each other to cause rounding errors.

**Comment by hdgarrood** [comment]
So apparently there are lots of things that can affect results, like rounding modes, compiler optimisation levels, and so on, but you have to compile the program again with different options or run it on a different machine or something like that for there to be a change in any of those (according to [this post](https://randomascii.wordpress.com/2013/07/16/floating-point-determinism/). It sounds like the failure @andyarvanitis described can happen randomly on different executions of the program on the same machine, without even recompiling, which leads me to suspect that it isn't floating point.

**Comment by jplatte** [comment]
> It sounds like the failure @andyarvanitis described can happen randomly on different executions of the program

Where did you read that? The floating point settings can be changed at runtime, but they are definitely not chosen at random. Also they should be the same on >99% of all systems by default, and you'll have to search a while to find a program that changes them (which I'm pretty sure would only affect that process and possibly child processes, but not the whole system).

**Comment by hdgarrood** [comment]
Yes - I wasn't suggesting that floating point settings can be chosen at random. It's the the infinite loop in the original issue description which can happen randomly.

**Comment by jplatte** [comment]
Well, why would it be random with the same floating point settings? Floating point operations can have rounding errors, but that doesn't mean they're non-deterministic. They just don't fulfill all the laws you'd want them to fulfill.

**Comment by hdgarrood** [comment]
Yes, this is exactly what I was getting at; this is why I suspect the underlying cause isn't a floating point error.

**Comment by jplatte** [comment]
Oh, that's what you meant by

> which leads me to suspect that it isn't floating point

Sorry for the confusion ^^°

**Comment by hdgarrood** [comment]
Haha, no worries. :)

**Comment by paf31** [comment]
The problem with exact numbers is that I'm not sure you can have both of a) express everything you want to be able to, and b) keep various operations like `==` total.

**Comment by andyarvanitis** [comment]
> It sounds like the failure @andyarvanitis described can happen randomly on different executions of the program on the same machine, without even recompiling, which leads me to suspect that it isn't floating point.

I didn't mean to imply that. My point was that if you change the computation (but keep the general program the same) the use of the equals operator could give you a different program result (and in this case an unsafe one).

**Comment by hdgarrood** [comment]
Ah, my bad. Just out of interest, do you know what changed to make this cause an infinite loop?

**Comment by andyarvanitis** [comment]
Yup, compiler optimizations off vs on, but only recently on top of a change in size of the value "box" class (nothing directly related to math operations). And even then, only on one compiler/OS combination. Before that, I wasn't aware of this test -- it had been merrily passing for the last year.

**Comment by andyarvanitis** [comment]
I should state: most languages (including even Haskell) don't offer any special protections for this. I just wanted to make sure to bring it up before simply offering to change the test (which I will gladly do).

**Comment by paf31** [comment]
Changing the test sounds good to me 👍
